src/main/java/io/vertx/ext/mail/MailMain.java

```diff
@@ -0,0 +1,613 @@
+
+  private void upgradeTLS() {
+    ns.upgradeToSsl(v -> {
+      log.debug("ssl started");
+      // capabilities may have changed, e.g.
+      // if a service only announces PLAIN/LOGIN
+      // on secure channel
+      ehloCmd();
+    });
+  }
+
+  private List<String> parseEhlo(String message) {
+    // parse ehlo and other multiline replies
+    List<String> v = new ArrayList<String>();
+
+    String resultCode = message.substring(0, 3);
+
+    for (String l : message.split("\n")) {
+      if (!l.startsWith(resultCode) || l.charAt(3) != '-' && l.charAt(3) != ' ') {
+        log.error("format error in multiline response");
+        throwAsyncResult("format error in multiline response");
+      } else {
+        v.add(l.substring(4));
+      }
+    }
+
+    return v;
+  }
+
+  private void authCmd() {
+    if (capaAuth.contains("CRAM-MD5")) {
+      write("AUTH CRAM-MD5", result -> {
+        String message = result.result();
+        log.debug("AUTH result: " + message);
+        cramMD5Step1(message.substring(4));
+      });
+    } else if (capaAuth.contains("PLAIN")) {
+      String authdata = base64("\0" + config.getUsername() + "\0" + config.getPassword());
+      write("AUTH PLAIN " + authdata, 10, result -> {
+        String message = result.result();
+        log.debug("AUTH result: " + message);
+        if (!message.toString().startsWith("2")) {
+          log.warn("authentication failed");
+          throwAsyncResult("authentication failed");
+        } else {
+          mailFromCmd();
+        }
+      });
+    } else if (capaAuth.contains("LOGIN")) {
+      write("AUTH LOGIN", result -> {
+        String message = result.result();
+        log.debug("AUTH result: " + message);
+        sendUsername();
+      });
+    } else {
+      log.warn("cannot find supported auth method");
+      throwAsyncResult("cannot find supported auth method");
+    }
+  }
+
+  private void cramMD5Step1(String string) {
+    String message = decodeb64(string);
+    log.debug("message " + message);
+    String reply = hmacMD5hex(message, config.getPassword());
+    write(base64(config.getUsername() + " " + reply), 0, result -> {
+      String message2 = result.result();
+      log.debug("AUTH step 2 result: " + message2);
+      cramMD5Step2(message2);
+    });
+  }
+
+  private String hmacMD5hex(String message, String pw) {
+    try {
+      SecretKey key = new SecretKeySpec(pw.getBytes("utf-8"), "HmacMD5");
+      Mac mac = Mac.getInstance(key.getAlgorithm());
+      mac.init(key);
+      return encodeHex(mac.doFinal(message.getBytes("utf-8")));
+    } catch (UnsupportedEncodingException | NoSuchAlgorithmException | InvalidKeyException e) {
+      // doesn't happen, auth will fail in that case
+      return "";
+    }
+  }
+
+  /**
+   * @param outBytes
+   * @return
+   */
+  private String encodeHex(byte[] bytes) {
+    StringBuilder sb = new StringBuilder(bytes.length * 2);
+    for (byte b : bytes) {
+      sb.append(String.format("%02x", b));
```

Comment:
I would recommend to use Integer.toHexString() with padding correction as String.format is expensive for just formatting a byte[] to hex string.